File: symphony/workflow.py
from __future__ import annotations

from pathlib import Path
from typing import Any

import yaml

from .errors import WorkflowError
from .models import WorkflowDefinition


def default_workflow_path(cwd: Path | None = None) -> Path:
    return (cwd or Path.cwd()) / "WORKFLOW.md"


def resolve_workflow_path(path: str | Path | None, cwd: Path | None = None) -> Path:
    selected = Path(path) if path is not None else default_workflow_path(cwd)
    return selected.expanduser().resolve(strict=False)


def load_workflow(path: str | Path | None = None, cwd: Path | None = None) -> WorkflowDefinition:
    workflow_path = resolve_workflow_path(path, cwd)
    try:
        raw = workflow_path.read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise WorkflowError("missing_workflow_file", f"workflow file not found: {workflow_path}", cause=exc) from exc
    except OSError as exc:
        raise WorkflowError("missing_workflow_file", f"workflow file cannot be read: {workflow_path}", cause=exc) from exc

    config: dict[str, Any]
    body: str
    if raw.startswith("---"):
        lines = raw.splitlines()
        closing_index: int | None = None
        for index, line in enumerate(lines[1:], start=1):
            if line.strip() == "---":
                closing_index = index
                break
        if closing_index is None:
            raise WorkflowError("workflow_parse_error", "YAML front matter is missing closing ---")
        front_matter = "\n".join(lines[1:closing_index])
        body = "\n".join(lines[closing_index + 1 :])
        try:
            parsed = yaml.safe_load(front_matter) if front_matter.strip() else {}
        except yaml.YAMLError as exc:
            raise WorkflowError("workflow_parse_error", f"invalid YAML front matter: {exc}", cause=exc) from exc
        if parsed is None:
            config = {}
        elif isinstance(parsed, dict):
            config = parsed
        else:
            raise WorkflowError("workflow_front_matter_not_a_map", "YAML front matter must decode to a map/object")
    else:
        config = {}
        body = raw

    try:
        mtime_ns = workflow_path.stat().st_mtime_ns
    except OSError:
        mtime_ns = None

    return WorkflowDefinition(config=config, prompt_template=body.strip(), path=workflow_path, mtime_ns=mtime_ns)

File: lib/symphony/workflow.ex
defmodule Symphony.Workflow do
  @moduledoc false

  alias Symphony.Error
  alias Symphony.Models.WorkflowDefinition

  def default_workflow_path(cwd \\ File.cwd!()) do
    Path.join(cwd, "WORKFLOW.md")
  end

  def resolve_workflow_path(path, cwd \\ File.cwd!())
  def resolve_workflow_path(nil, cwd), do: cwd |> default_workflow_path() |> Path.expand()
  def resolve_workflow_path(path, _cwd), do: path |> to_string() |> Path.expand()

  def load_workflow(path \\ nil, cwd \\ File.cwd!()) do
    workflow_path = resolve_workflow_path(path, cwd)

    raw =
      case File.read(workflow_path) do
        {:ok, body} ->
          body

        {:error, reason} ->
          raise Error,
            code: :missing_workflow_file,
            message: "workflow file cannot be read: #{workflow_path}",
            cause: reason
      end

    {config, body} = parse_workflow(raw)

    %WorkflowDefinition{
      config: config,
      prompt_template: String.trim(body),
      path: workflow_path,
      mtime_ns: mtime(workflow_path)
    }
  end

  defp parse_workflow("---" <> _ = raw) do
    lines = String.split(raw, ~r/\R/, trim: false)

    closing_index =
      lines
      |> Enum.drop(1)
      |> Enum.find_index(&(String.trim(&1) == "---"))

    if is_nil(closing_index) do
      raise Error,
        code: :workflow_parse_error,
        message: "YAML front matter is missing closing ---"
    end

    closing_index = closing_index + 1
    front_matter = lines |> Enum.slice(1, closing_index - 1) |> Enum.join("\n")
    body = lines |> Enum.slice((closing_index + 1)..-1//1) |> Enum.join("\n")

    parsed =
      if String.trim(front_matter) == "" do
        %{}
      else
        case YamlElixir.read_from_string(front_matter) do
          {:ok, value} ->
            value

          {:error, reason} ->
            raise Error,
              code: :workflow_parse_error,
              message: "invalid YAML front matter: #{inspect(reason)}"

          value ->
            value
        end
      end

    cond do
      is_nil(parsed) ->
        {%{}, body}

      is_map(parsed) ->
        {parsed, body}

      true ->
        raise Error,
          code: :workflow_front_matter_not_a_map,
          message: "YAML front matter must decode to a map/object"
    end
  end

  defp parse_workflow(raw), do: {%{}, raw}

  def mtime(path) do
    case File.stat(path, time: :nanosecond) do
      {:ok, stat} -> stat.mtime
      _ -> nil
    end
  end
end
